tutorial/README.md

```diff
@@ -0,0 +1,413 @@
+| 112232|128585|   30390|  44375|    914|  2191|      1|   126.00|     18.90|
+|  97099|110826|   24730|  32758|   3331|  2190|      1|   319.00|     47.85|
+|  74902| 85497|   45049|   2318|   7547|  2190|      2|   554.00|     83.10|
++-------+------+--------+-------+-------+------+-------+---------+----------+
+only showing top 20 rows
+```
+
+### Under the hood - Load Function ###
+
+In this section we will take a peek inside spark-redshift to understand how exactly the LOAD function works. Specifically we will look at how `event` table was read into a DataFrame. The LOAD is a two step process
+
+1. [UNLOAD](http://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD.html) data from the Redshift table to S3
+2. Consume the unloaded files in S3 via a custom InputFormat into an RDD which is then wrapped as a DataFrame instance using the schema obtained from Redshift.
+
+#### UNLOAD Redshift to S3 ####
+
+Figure 1 shows the the actions required to perform this step.
+
+![](images/loadunloadstep.gif)
+
+Figure 1 : UNLOAD action
+
+First the Spark Driver communicates with the Redshift Leader node to obtain the schema of the table (or query) requested. The attribute `override lazy val schema: StructType` in the class `com.databricks.spark.redshift.RedshiftRelation` will obtain the schema on demand by invoking the method `resolveTable` of the class `com.databricks.spark.redshift.JDBCWrapper`. The `JDBCWrapper` class is responsible for fetching the schema from the Redshift Leader.   
+
+Next a Redshift [UNLOAD](http://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD.html) query is created using the schema information obtained. The UNLOAD command unloads each slice into a s3 folder (22c365b4-13cb-40fd-b4d6-d0ac5d426551) created in the temporary s3 location (s3n://spark-redshift/temp/) provided by the user. Each file contains a row per line and each column of the row is pipe (|) delimited. This process occurs in parallel for each slice. The spark-redshift library achieves its high performance through this mechanism.
+
+#### Read UNLOAD'ed S3 files into a DataFrame instance####
+
+Figure 2 shows how the files unloaded in S3 are consumed to form a DataFrame instance
+
+![](images/loadreadstep.gif)
+Figure 2 : Read from S3 into a DataFrame  
+
+Once the files are written to S3, a custom InputFormat implemented in the class `com.databricks.spark.redshift.RedshiftInputFormat` is used to consume the files in parallel. This class is similar to the standard and well known `TextInputFormat` where the key is the byte offset of the start of each line in the file. The value class however, is of type `Array[String]` (unlike `TextInputFormat`where it is `Text`). The value instance is created by splitting the lines using the default delimiter |. The `RedshiftInputFormat` consumes the S3 files line by line to produce an `RDD`. The schema obtained earlier is applied on this `RDD` to generate an instance of a `DataFrame`.
```

Comment:
This first sentence would be an awesome opportunity to use more commas, since "implemented in the class `blahblahblah`" is kind of a parenthetical.

Similarly, " using the default delimiter |" needs either backticks around the pipe character, a comma before it, or both.

In the last sentence, you say "an RDD" or "this RDD", but then use the slightly-awkward "an instance of a DataFrame." Why not just say "a DataFrame"?